### PR TITLE
Drop support for 32 bit systems/PHP

### PR DESCRIPTION
**Pull request by dktapps**
long overdue

### CODE REVIEW

**Comment by TheDiamondYT1** [comment]
Hey, i've had a great idea! Lets drop support for binding to custom ports! :D

**Comment by Muqsit** [comment]
Who even uses 32 bit lol

**Comment by JackNoordhuis** [comment]
About time this happened \o/

**Comment by dktapps** [comment]
You've had over 8 months deprecation warning (since 0.16).

**Comment by HimbeersaftLP** [comment]
> Why dont we just move to horrible piece of **** called Tesseract. At least it will work on my computer.

Good point, I think it isn't a good idea to completely cut the support, because then even more poeple will move to spoons because they say exactly that (I won't, all PCs I run pmmp on are 64 bit)

**Comment by AppleDevelops** [comment]
Removing 32 bit support is for the best though, it's slow and is starting to become obsolete.

**Comment by thebigsmileXD** [comment]
I use 2 32x machines too, both just for test servers though. I just heard several hosting sites only support x32 systems? The hell?

Though the move is needed. Several issues can be fixed by that. <3 x64

**Comment by AppleDevelops** [comment]
Aren't hosting sites supposed to be **powerful** and **good**?

**Comment by TheDiamondYT1** [comment]
@AppleDevelops

> starting to become obselete

"Starting to"

If you were apple, you could argue that headphone jacks are "starting" to become obselete. Others may not agree.

**Comment by dktapps** [comment]
To clarify the reasoning for this:

This is not an on-a-whim change, it's a technical decision based on the fact that
- 32-bit systems have become enough of a pain in the ass to cause problems with development.
- Many things don't work properly on 32-bit servers (noticed you can't get rid of the oxygen bar on 32-bit now?), primarily things which need 64-bit integers
- 32-bit servers are primarily made up of old crap computers and mobile phones, neither of which are suitable for running a server on.
- 32-bit servers cannot perform as well as a 64-bit server can, primarily due to the need to use strings for storing hashes and long integers on 32-bit platforms and constantly casting back and forth.

The TL;DR of this is that the lack of an int64 type on 32-bit PHP is making my job hard work, and I'm sick of it. It doesn't make sense to run a server on 32-bit. If you want to blame someone, you can blame me, or blame shoghicp for choosing PHP (if this was Java we wouldn't have any such problem).

**Comment by Driesboy** [comment]
i use 32 bit

**Comment by TheDiamondYT1** [comment]
Ok then lets purge the repo and start java

**Comment by Muqsit** [comment]
For everyone saying "I use 32 bit"; If you don't see it already, it's much better to upgrade to 64 bit servers. And what the fa? "32 bit servers for testing", yes its good but by saying that you mean you are using 64 bit servers for production right? You'll get much better results if you use the same bit for both.

**Comment by TheDiamondYT1** [comment]
@Muqsit Okay then please send me the money and i will

**Comment by Muqsit** [comment]
@TheDiamondYT1 Its best I don't waste my money on 32-bit fans. About time bro! Or don't host PocketMine :P

**Comment by TheDiamondYT1** [comment]
@dktapps Cant you *at least* add a startup argument to allow running the server on 32-bit (so we can still test our plugins on our shitty 32-bit PCs) but make it very clear that alot of stuff wont work

**Comment by TheDiamondYT1** [comment]
@Muqsit link?

**Comment by Muqsit** [comment]
Deleting this comment in 2 minutes due to advertisement policies.
https://www.quora.com/Which-companies-offer-free-VPS-hosting

**Comment by dktapps** [comment]
@TheDiamondYT1 With the move to strict types 32-bit will just crash everywhere. It might start but you wouldn't be able to join it.

**Comment by TheDiamondYT1** [comment]
@dktapps I'll just do what i always do and comment out the `exit()`

**Comment by xinghao2003** [comment]
Although using 64bit is better, but many user computer still in 32bit maybe can create a new repo that about 32bit PocketMine

**Comment by TheDiamondYT1** [comment]
@xinghao2003 That will **never** happen.

**Comment by xinghao2003** [comment]
my computer is 64bit, but i always test my plugins in my android 32bit so now i cant test anymore in my phone

**Comment by TheDiamondYT1** [comment]
@xinghao2003 You can, just use the latest build before the binary strict types commit and you have to comment out the `exit()` code if you dont have the zip exstension installed (since php binary on android doesnt have it)

**Comment by Qimie** [comment]
Hey. For all those PC users that their architecture is 32 bit. U can use
Bluestacks to run Pocketmine I believe.

**Comment by jasonwynn10** [comment]
PMMP was never officially supported on mobile devices in the first place. I agree with the dropped support of x32

**Comment by Qimie** [comment]
Then why do we have the Pocketmine app. -_-

**Comment by jasonwynn10** [comment]
@Qimie the PocketMine app was made by PocketMine, not the PMMP team. -_-

**Comment by KnownUnown** [comment]
Regardless, this is all beyond the point. IIRC Bluestacks inherits the host architecture, therefore, running from the PocketMine Team-made app on BlueStacks, whether supported or not, is not a valid workaround. It's pretty convoluted anyways.

**Comment by Qimie** [comment]
Ok. Whatever u say. Ur better at coding than me. I cant fight yah. XD

**Comment by Frago9876543210** [comment]
This a good idea!

**Comment by SOF3** [comment]
@jasonwynn10 PMMP and PocketMine are identical. The PMMP Team is the PockektMine Team, just that some people became inactive and some new people joined.

**Comment by TheDiamondYT1** [comment]
USB-C for headphones is the dumbest fucking idea, ever.

**Comment by TheDiamondYT1** [comment]
@SOF3 okay then lets purge the repo and piggyback off MiNET :)

**Comment by Awzaw** [comment]
> okay then lets purge the repo and piggyback off MiNET :)
I assume that's a joke, because as good as MiNET is, it won't work on Windows XP, Linux or OSX.

**Comment by dschwartz783** [comment]
I can say with almost complete certainty that almost everyone uses 64-bit servers these days (no, your phone doesn't count). 64-bit processors have been shipped by Intel since 2007!!! We're talking a decade here, folks. Heck, the dirt-cheap raspberry pi is 64-bit now, which tells you something. If you have a computer that doesn't support 64-bit, it was either very cheap when you bought it (Intel Atom), or it's very very old, and needed to be replaced 5 years ago. It certainly isn't something you should be hosting a server on. The few people that need to work in a 32-bit environment can either pick one of the various crappy spoons available, or switch to something architecture-independent like Nukkit. Or better yet, upgrade your junk. Use one of the 64-bit linux distributions available for the Raspberry Pi, and you'll be good to go.

**Comment by thebigsmileXD** [comment]
Sorry if annoying, but can you move this to the forums please? Thanks ;) have a nice day.

**Comment by dktapps** [comment]
> no, your phone doesn't count

even my phones are all 64-bit, bar a Sony Xperia Z2 (2014) :P

**Comment by xxFlare** [comment]
This change is plausible.

You may want to add a note on the issue tracker that reminds people 32bit systems are not supported. Though people will probably still open issues regardless.

**Comment by MattWAnderson** [comment]
+1 - If you are running 32 bit, you should of upgraded years ago.

**Comment by ParkChanSol** [comment]
Oops.. My computer is 32 bit.  ((ㅇㅅㅇ));;

**Comment by SOF3** [comment]
This RFC requests for comments on dropping 32-bit support. This is not a poll asking what architecture your machine uses. Commenting to tell whether your computer is 32-bit is irrelevant to this pull request if you aren't comparing the advantages and disadvantages of this RFC, because nobody cares if your very own computer is 32-bit.

**Comment by TheDiamondYT1** [comment]
@SOF3 I use 32-bit.

**Comment by IrishPacks** [comment]
> (slimmer design and better audio)

Better audio? No, no, no, no, and no.

**Comment by AppleDevelops** [comment]
@pmmp-bot Two different things, dropping 32 bit support isn't being a racist, while assuming all people born are white is racist.

@dktapps It wouldn't be a bad idea to lock this.

**Comment by dschwartz783** [comment]
@pmmp-bot You could probably find someone with a 16-bit computer too. Is it racist that NOBODY supports them? 32-bit computers have finally become a small minority. This project is developed by people in their free time. They choose how much work they want to take on themselves, not the users. If you don't like it, feel free to move on.

**Comment by Muqsit** [comment]
@pmmp-bot Yes you are right. We cant expect 95% of the people to be having 64-bit operating systems. It should have actually been 99.9% considering barely anyone uses a decade old PC, be it even for testing.

**Comment by dktapps** [comment]
Just in case anyone was wondering, @pmmp-bot has nothing to do with PMMP.

**Comment by dktapps** [comment]
@pmmp-bot that comparison would make a tiny bit of sense if non-white people were obsolete... they aren't.

I'm terribly sorry if you happen to be among the 0.1% who happens to use 32-bit for whatever reason... but our lives don't revolve around you. Fuck your discrimination suggestions. I'm making an informed decision. What are you gonna do next? Razz Microsoft for dropping Windows XP support? 😱 

You can't make a pair of jeans that fits everyone. Which is why you cater for the majority to improve the majority's experience.

For far too long people have taken PocketMine-MP and its developers for granted. As the primary active developer I am making a decision to drop support for obsolete systems to make my life easier and to improve things for the many, not for the few.

**Comment by dktapps** [comment]
The only reason we've supported 32-bit this long is because I put a hard weekend of effort into fixing varint64 for 32-bit. I should have dropped support way back in October at 0.16 when I first said this and saved myself the time and effort. Did it benefit me at all? It did not, I did it out of the goodness of my heart.

The tone of the comments coming out here is one of "I don't care as long as it works for me". That's the attitude in general as I see it. Yet the second I do something that actually does affect you, instead of behaving professionally you turn round and attack me for something I was never obliged to do in the first place. I think that is low and selfish. As I said - you take us for granted.

**Comment by xxFlare** [comment]
@pmmp-bot 

You just compared a server to a skin color... what?

Also, Apple just dropped 32bit app support yesterday with the launch of IOS 11 beta. Upgrade your device.

**Comment by dktapps** [comment]
It seems I have a majority vote on this... I'd like to get this done as soon as possible.

**Comment by dktapps** [comment]
Don't mind Travis... PECL is down so all builds are erroring.

**Comment by dktapps** [comment]
Oh, and in case anyone misunderstood... This does not put any requirements on clients. 

This only makes it a requirement that the machine/device **hosting** the server must be 64-bit and be running on 64-bit PHP. This change does not affect clients in any way.

**Comment by MattWAnderson** [comment]
I agree, I'd suggest just merging it and get the bandaid pulled off quickly, rather than it being talked to death here with the same end result.

**Comment by TheDiamondYT1** [comment]
# Fuck my life

**Comment by xxFlare** [comment]
@TheDiamondYT1 Is it that hard to buy a new computer?

**Comment by Driesboy** [comment]
yep its hard

**Comment by dktapps** [comment]
There's nothing stopping you cloning the repository and reverting commit 4765242 to run the server. But do note that we will provide no support whatsoever for servers running on 32-bit from this point onwards. If you insist on using outdated platforms, be prepared to fix them yourselves, because we aren't going to do it anymore.

**Comment by TheDiamondYT1** [comment]
@xxFlare do you have the money?

**Comment by HimbeersaftLP** [comment]
> @xxFlare do you have the money? Im 14.

Just save sum money, cheap 64 bit laptops can be bought for under 300$, even cheaper used...

**Comment by xxFlare** [comment]
@HimbeersaftLP @TheDiamondYT1  You don't even have to save. If you bought some bitcoins in October (which alternatively was a ton of time to save up fiat cash) when dktapps first announced the drop in support, your small investment would have multiplied 2.5x.

**Comment by dschwartz783** [comment]
@TheDiamondYT1  Raspberry Pi, bro. If you're planning on becoming a serious developer, you should get familiar with Linux anyway. Try the SuSE linux x64 release for it. In a week or two, Fedora will have an x64 release for it too.

**Comment by Qimie** [comment]
My 32 bit phone can still use pmmp. Lol.

**Comment by dschwartz783** [comment]
@Qimie Yeah, and I can still make telephone calls with two cups and a string.

**Comment by Qimie** [comment]
Suck it. Im done.

**Comment by TheDiamondYT1** [comment]
@dschwartz783 I'm familiar with Linux.

**Comment by jojoe77777** [comment]
This pull request conversation is not supposed to be about if @TheDiamondYT1 can afford a 64bit computer or not.

**Comment by dschwartz783** [comment]
@jojoe77777 This pull request should be locked. It isn't even a PR anymore, really. It's been merged.
